src/toil/jobStores/fileJobStore.py

```diff
@@ -370,35 +371,33 @@ def readFile(self, jobStoreFileID, localFilePath, symlink=False):
         self._checkJobStoreFileID(jobStoreFileID)
         jobStoreFilePath = self._getAbsPath(jobStoreFileID)
         localDirPath = os.path.dirname(localFilePath)
-        # If local file would end up on same file system as the one hosting this job store ...
-        if os.stat(jobStoreFilePath).st_dev == os.stat(localDirPath).st_dev:
-            # ... we can hard-link the file, ...
-            if symlink:
-                try:
+        if symlink or self.linkImports:
```

Comment:
The [`readGlobalFile` implementation in the `NonCachingFileStore`](https://github.com/DataBiosphere/toil/blob/37b55569122e7d3f5e260d149bf4d62ca71187f1/src/toil/fileStore.py#L1665-L1675) assumes that if you don't pass `symlink=True` to `readFile` you will always get a fresh, mutable copy, which is how it manages to ignore its own `mutable` parameter. If we change things so that that won't always be the case (i.e. if the user passes `--linkImports`), we need to change the `readGlobalFile` implementation to deal with that.

Given the current code here, which will symlink or hard link even if the `readFile` caller passes `symlink=False`, the only way to get a mutable copy to fulfill a user request for one would be to make a copy ourselves in the  `NonCachingFileStore`.

Also, since this PR touches the semantics of the `symlink` kwarg, which doesn't have a description in any of the docstrings for `readFile` or `readGlobalFile`, this PR should define the semantics for that parameter and document them in the docstrings of those functions. Since the option is now going to result in either a symlink *or* a hardlink, as appropriate, the name should probably be changed to something like `allowLink`. It could also be removed altogether, in favor of bringing the `mutable` option down into `readFile`; maybe @DailyDreaming can comment on how `symlink` and `mutable` need to differ.

Reply:
I agree that `symlink` and `mutable` seem to serve similar purposes in `readGlobalFile` and `mutable` alone could be brought in and used instead of `symlink`, though perhaps a clearer name would be nice.

Reply:
I think that `mutable` is not currently implemented right. The code as it is now will create a hard link even if you pass `mutable=True`, if it thinks everything is on the same device. And the official docs for `mutable` are missing; the base class refers you to the `CachingFileStore` docstring, which doesn't exist.